### ISSUE TITLE
Feature Request: Allow for specific internal IPs

### ISSUE THREAD

**Issue by HybridSarcasm** [mkg20001/nat-puncher#2]
Great project. One request: can you provide a way to create our own array of internal IPs? The list you've statically defined in utils.js just happens to be completely inapplicable to my internal network. Because of this, the [Ring plugin for Homebridge](https://github.com/dgreif/ring) cannot properly create the UDP port mapping required for live video streaming from my Ring Video doorbell.

Please let me know if there's any other information I can provide for this feature request. Thanks!

**Comment by mkg20001** [comment]
I've added a feature that now makes it guess the gateway base on the client ip, but that will only guess gateways that end with a one (in my case for example my machine has the address `192.168.77.120` so it guesses `192.168.77.1`)

https://github.com/mkg20001/nat-puncher/commit/bf0381b7bd3f1f9aaae664d2d10e881bc8a45eed#diff-2b4ca49d4bb0a774c4d4c1672d7aa781R6-R20

That fails for all .254 gateways

Should I add those, too?

Should I add an option for the user to add their own gateways as well?

**Comment by HybridSarcasm** [comment]
Hi @mkg20001. Thank so much! Perhaps my use of .254 as the gateway is a bit uncommon. It may not be worth the extra broadcast traffic. But, allowing the use of a specific gateway IP would be fantastic!

**Comment by dgreif** [comment]
@mkg20001 I'm the maintainer of the `homebridge-ring` plugin that @SeminolesFan78 is using.  I think your last change makes a ton of sense and is a good addition!  For @SeminolesFan78's specific case, I think it might be worth it to just pass in a specific gateway address.  I'm happy to add the config option on my end and pass it along to your library.  Thanks for the quick response time and a great tool!

**Comment by dgreif** [comment]
@mkg20001 another option would be to use something like https://github.com/silverwind/default-gateway to grab the default gateway from the OS.  Maybe running that inside a try/catch like your other other addition.  @SeminolesFan78 would the default gateway correctly point to .254 in your case?

**Comment by HybridSarcasm** [comment]
@dgreif You got it. I use .254 as the gateway for all the subnets on my home network.

**Comment by mkg20001** [comment]
I've released a fix thats include default-gateway and my own gateway detection code

default-gateway is an optional dependency, so make sure to have it installed for that feature to work.

As for the option: I have no clue how it would be best implemented, feel free to PR

**Comment by dgreif** [comment]
@mkg20001 the `default-gateway` update worked perfectly!  I am now including `default-gateway` as a dependency in `hombridge-ring` so that it's always present for my users.  Several users that had issues opening ports previously are now able to do so without issue 🎉 .  Thank you so much for getting this fix in!

From an implementation standpoint, two thoughts:
 * Would you consider adding this as a true dependency instead of an optional dependency?  I think this could be a major headache for other projects using `nat-puncher` if they don't realize that requests aren't being sent to their default gateway.
 * Would it be possible to use the async interface with `default-gateway`?  It runs quickly on my machine (macbook pro), but I'd hate to block the thread waiting for it to run if it's slow on other platforms

As for the option to specify which ip to use, I think this is a lot less useful now that default gateway is being used.  I think situations where you use anything _except_ default gateway would be rare.

**Comment by mkg20001** [comment]
> Would it be possible to use the async interface with default-gateway?

I don't know if it would be possibly without rewriting things too much.

> Would you consider adding this as a true dependency instead of an optional dependency? 

Sure, can do

**Comment by mkg20001** [comment]
Released a new version

**Comment by dgreif** [comment]
Awesome, thanks @mkg20001!  I think this issue has been resolved now 🎉.  I do have a question for you tho, as you seem to know more about nat-punching than I do.  This plugin (along with sending an empty udp packet _out_ through the port) seems to work in 99% of my user's setups.  I still see inbound UDP packets being blocked when on a VPN and some other odd complex networks for some reason.  Any idea how to do a bullet-proof punch to open up the port for UDP to flow in?  If you don't know of anything beyond what this library is already doing, then no worries.  Thanks!

**Comment by mkg20001** [comment]
I think when using a VPN, packets routed via the "real" network interface might be blocked by some clients. Especially UDP as it happens to be the way DNS packets could leak.

**Comment by blop** [comment]
Currently nat-puncher assumes that we always add a mapping to our own internal IP address (which is not the case for me :p).

It'd be nice to have the internalIPAddress as a addMapping() parameter (could be optional).
In the meantime I'll look to override the getPrivateIps() function.

**Comment by Tesla-P100FX** [comment]
I would love to get a little help with this.   Plug in is awesome btw.  Thank you! 

I read this thread but still not sure how to resolve that i cant stream when i have a VPN active.    Ring Native client works, the plug in works with the VPN turned off.  I get snapshots but it wont stream with it on.  Log attached
[Ring Log File.zip](https://github.com/mkg20001/nat-puncher/files/4978689/Ring.Log.File.zip)
.

**Comment by donavanbecker** [comment]
@Tesla-P100FX, I don't believe the ring plugin is using this anymore.

**Comment by dgreif** [comment]
That's correct, the ring plugin now takes a different approach to getting the ports open properly, and this library is no longer used in the process. @Tesla-P100FX I have an update that I will release in the next couple days. I suggest you try that once it's released. @mkg20001 You can probably course this issue since it isn't needed for ring anymore

**Comment by Tesla-P100FX** [comment]
Great thank you. I will look forward to getting the update so that'll work properly. Again nice job on this plug-in￼

**Comment by blop** [comment]
I think the use case to be able to specify the internal IP to use is not limited to the "ring" project.
I suggest we keep this ticket open ;-) 
I have a working patch for it but didn’t clean it up enough to make a PR yet.

**Comment by Tesla-P100FX** [comment]
My Homebridge device turns non responsive with the current Ring Plugin installed with 9 cameras.    Hopefully the new one solves for this.

**Comment by dgreif** [comment]
@Tesla-P100FX I'm not sure if you are aware, but you are posting these comments on the `mkg20001/nat-puncher` repo, not the `dgreif/ring` repo.  This is not the appropriate place for these comments.
